### PR TITLE
enable android build

### LINKED ISSUE #271: deploy on ARM device(Android)
**Is your feature request related to a problem? Please describe.**
         I want to deploy onnxruntime on android device, but I can't make progress because of the lack of 
   related guide from github, only find "Cross compiling on Linux" ,but this is not enough for me, would 
   you please update the ARM build guide in more detail?
**System information**
- ONNX Runtime version (you are using):
   0.1.5
**Describe the solution you'd like**
   an ARM build guide(Linux) in detail 

**Describe alternatives you've considered**
   N/A

**Additional context**
   NA

### PR DESCRIPTION
**Pull request by daquexian**
Fix #271, tested with android ndk r19b.

I don't use build.sh but only cmake

```
cmake -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake -DANDROID_CPP_FEATURES=exceptions -DANDROID_PLATFORM=android-28 -DANDROID_ABI=arm64-v8a -DCMAKE_BUILD_TYPE=Release -Donnxruntime_CROSS_COMPILING=ON -Donnxruntime_BUILD_x86=OFF -DONNX_CUSTOM_PROTOC_EXECUTABLE=/usr/bin/protoc ../cmake 
```

My changes:
1. Android NDK (at least for r19b) always output `x86_64-unknown-linux-gnu` for `-dumpmachine` on my PC, so the arch should be determined by the cmake official way
2. <del>Android NDK doesn't support filesystem so far (https://github.com/android-ndk/ndk/issues/609), so I set the two flags explicitly for Android. What's more, the `header_files_test` should be disabled if filesystem is not supported.</del> Filesystem is not needed since fb2a44f6423ac5d4f8af9105af82f00c99ae3a46.

------------------

The result of `onnxruntime_test_all` on Android:

> [----------] Global test environment tear-down                                                                                                           
[==========] 957 tests from 104 test cases ran. (3713 ms total)                                                                                          
[  PASSED  ] 918 tests.                                                                                                                                  
[  FAILED  ] 39 tests, listed below:                                                                                                                     
[  FAILED  ] GraphTransformationTests.IdentityElimination                                                                                                
[  FAILED  ] GraphTransformationTests.SliceElimination                                                                                                   
[  FAILED  ] GraphTransformationTests.ConstantFolding1                                                                                                   
[  FAILED  ] GraphTransformationTests.FuseConvBNMulAddUnsqueeze                                                                                          
[  FAILED  ] GraphTransformationTests.FuseConvActivation                                                                                                 
[  FAILED  ] GraphTransformationTests.FuseConvBNNoBias                                                                                                   
[  FAILED  ] GraphTransformationTests.FuseConvMulNoBias                                                                                                  
[  FAILED  ] GraphTransformationTests.FuseConvAddNoBias                                                                                                  
[  FAILED  ] GraphTransformationTests.FuseConvBNMulAddUnsqueezeNoBias                                                                                    
[  FAILED  ] GraphTransformationTests.FuseConvAddMul3D                                                                                                   
[  FAILED  ] GraphTransformationTests.MatMulAddFusion_two_input                                                                                          
[  FAILED  ] GraphTransformationTests.MatMulAddFusion_three_input                                                                                        
[  FAILED  ] GraphTransformationTests.Gemm_Relu_three_input                                                                                              
[  FAILED  ] GraphTransformationTests.FuseConvBnAddMulFloat16                                                                                            
[  FAILED  ] Float16_Tests.Mul_16_Test                                                                                                                   
[  FAILED  ] InferenceSessionTests.NoTimeout                                                                                                             
[  FAILED  ] InferenceSessionTests.DisableCPUArena                                                                                                       
[  FAILED  ] InferenceSessionTests.CheckRunLogger                                                                                                        
[  FAILED  ] InferenceSessionTests.CheckRunProfilerWithSessionOptions                                                                                    
[  FAILED  ] InferenceSessionTests.CheckRunProfilerWithStartProfile                                                                                      
[  FAILED  ] InferenceSessionTests.MultipleSessionsNoTimeout                                                                                             
[  FAILED  ] InferenceSessionTests.PreAllocateOutputVector                                                                                               
[  FAILED  ] InferenceSessionTests.ConfigureVerbosityLevel                                                                                               
[  FAILED  ] InferenceSessionTests.TestWithIstream                                                                                                       
[  FAILED  ] InferenceSessionTests.TestRegisterExecutionProvider                                                                                         
[  FAILED  ] InferenceSessionTests.InvalidInputTypeOfTensorElement                                                                                       
[  FAILED  ] InferenceSessionTests.ModelWithoutOpset                                                                                                     
[  FAILED  ] InferenceSessionTests.TestTruncatedSequence                                                                                                 
[  FAILED  ] InferenceSessionTests.TestCopyToFromDevices                                                                                                 
[  FAILED  ] InferenceSessionTests.TestL1Transformers
[  FAILED  ] InferenceSessionTests.TestL1AndL2Transformers
[  FAILED  ] InferenceSessionTests.TestCustomTransformers                                                                                                
[  FAILED  ] InferenceSessionTests.DisableAllTransformers                                                                                                
[  FAILED  ] CustomKernelTests.CustomKernelWithBuildInSchema                                                                                             
[  FAILED  ] CustomKernelTests.CustomKernelWithCustomSchema
[  FAILED  ] CustomKernelTests.CustomKernelWithOptionalOutput                                                                                            
[  FAILED  ] MemcpyTest.copy1
[  FAILED  ] Random.MultinomialGoodCase                                                                                                                  
[  FAILED  ] Random.MultinomialDefaultDType

### CODE REVIEW

**Comment by daquexian** [comment]
Note: The test compiles for android only if I add
```
find_library(log-lib log)
target_link_libraries(libprotobuf-lite ${log-lib})
```
in `libprotobuf-lite.cmake` due to https://github.com/protocolbuffers/protobuf/issues/2719

However, `libprotobuf-lite.cmake` is from protobuf repo. So what is the elegant way to make this edition?

**Comment by snnn** [comment]
libprotobuf-lite is a static lib, you don't need to call target_link_libraries on it. Therefore, I don't think you need to change protobuf's source code.

File: cmake/onnxruntime_unittests.cmake [hunk 1]
```diff
@@ -228,6 +233,10 @@ endif()
 
 file(GLOB onnxruntime_test_framework_src ${onnxruntime_test_framework_src_patterns})
 
+if(NOT (HAS_FILESYSTEM_H OR HAS_EXPERIMENTAL_FILESYSTEM_H))
```
Comment (snnn) [inline cmake/onnxruntime_unittests.cmake]: Let me submit a PR that totally remove header_files_test.cc. Then you don't need to change this cmake file.

**Comment by daquexian** [comment]
@snnn  Thanks for your review. Even though protobuf-lite is a static lib, the compiler flag will be spread to the executable by cmake.

And what is your opinion on https://github.com/protocolbuffers/protobuf/issues/2719? `-llog` should be added somewhere to solve the undefined reference.

**Comment by daquexian** [comment]
@snnn I have rebased this branch on master. However I believe `-llog` should be added somewhere to solve the undefined reference caused by https://github.com/protocolbuffers/protobuf/issues/2719

**Comment by snnn** [comment]
In CMakeLists.txt, you may add "log" to onnxruntime_EXTERNAL_LIBRARIES.

**Comment by snnn** [comment]
Hi @manashgoswami, could you find someone verifying this PR? Thanks

**Comment by daquexian** [comment]
@snnn Thanks! I have updated my PR according to your advice. Now the tests compiles

**Comment by snnn** [comment]
Are you doing cross-compiling or native ?

**Comment by daquexian** [comment]
@snnn cross-compiling via android ndk.

**Comment by snnn** [comment]
>  the arch should be determined by the cmake official way

Any suggestion?

**Comment by daquexian** [comment]
@snnn My PR determines the architecture by CMAKE_ANDROID_ARCH_ABI, which is the variable set by CMake or the toolchain file. Please see the [cmake manual](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android) about [CMAKE_ANDROID_ARCH_ABI](https://cmake.org/cmake/help/latest/variable/CMAKE_ANDROID_ARCH_ABI.html#variable:CMAKE_ANDROID_ARCH_ABI).

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -581,6 +581,10 @@ else()
   set(onnxruntime_EXTERNAL_LIBRARIES_DEBUG ${onnxruntime_EXTERNAL_LIBRARIES})
 endif()
 
+if (CMAKE_SYSTEM_NAME STREQUAL "Android")
+  list(APPEND onnxruntime_EXTERNAL_LIBRARIES log)
```
Comment (yufenglee) [inline cmake/CMakeLists.txt]: What's this library for?

**Comment by daquexian** [comment]
Hi @snnn @yufenglee , is there still something I need to do for this PR? :)

**Comment by pranavsharma** [comment]
@daquexian can you please check-in a CI yaml file to ensure future changes don't break the Android build?

**Comment by daquexian** [comment]
@pranavsharma Hi, I have uploaded the new ci pipeline, however, I haven't tested it and I don't know how to enable it.

I build binary directly by cmake instead of build.py, so the python part is not needed. If you approve this way, I will remove all the code about python in android pipeline.

**Comment by daquexian** [comment]
I have tested this PR using my own azure account and verified the CI works(log: https://dev.azure.com/daquexian/onnxruntime/_build/results?buildId=14 for commit 1acfb67. I have reverted two commits which are only for my own azure pipeline). Now it can be tested in this Microsoft repo as long as an admin starts the CI. @linkerzhang @snnn @pranavsharma 

And @linkerzhang , could I get the privilege to start the CI of this repo?

**Comment by daquexian** [comment]
@pranavsharma Thanks! Could you please enable the new Android CI for this PR? The yaml is tools/ci_build/github/azure-pipelines/android-arm64-crosscompile-ci-pipeline.yml

**Comment by daquexian** [comment]
@linkerzhang Thanks!

**Pull request merged**